src/sources/exec.rs

```diff
@@ -0,0 +1,843 @@
+impl SourceConfig for ExecConfig {
+    async fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<super::Source> {
+        match self.mode.clone() {
+            Mode::Scheduled { exec_interval_secs } => {
+                run_scheduled(self.clone(), exec_interval_secs, shutdown, out)
+            }
+            Mode::Streaming {
+                respawn_on_exit,
+                respawn_interval_secs,
+            } => run_streaming(
+                self.clone(),
+                respawn_on_exit,
+                respawn_interval_secs,
+                shutdown,
+                out,
+            ),
+        }
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        EXEC
+    }
+}
+
+pub fn run_scheduled(
+    config: ExecConfig,
+    exec_interval_secs: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    Ok(Box::pin(async move {
+        info!("Starting scheduled exec runs.");
+        let schedule = time::Duration::from_secs(exec_interval_secs);
+        let mut interval = time::interval(schedule).take_until(shutdown.clone());
+
+        while interval.next().await.is_some() {
+            // Mark the start time just before spawning the process as
+            // this seems to be the best approximation of exec duration
+            let now = Instant::now();
+
+            // Wait for our task to finish, wrapping it in a timeout
+            let timeout = tokio::time::timeout(
+                schedule,
+                run_command(config.clone(), shutdown.clone(), out.clone()),
+            );
+
+            let timeout_result = timeout.await;
+
+            match timeout_result {
+                Ok(output) => {
+                    if let Err(command_error) = output {
+                        emit!(ExecFailed {
+                            command: config.command.as_str(),
+                            error: command_error,
+                        });
+                    }
+                }
+                Err(_) => {
+                    emit!(ExecTimeout {
+                        command: config.command.as_str(),
+                        elapsed_seconds: now.elapsed().as_secs(),
+                    });
+                }
+            }
+        }
+
+        info!("Finished scheduled exec runs.");
+        Ok(())
+    }))
+}
+
+pub fn run_streaming(
+    config: ExecConfig,
+    respawn_on_exit: bool,
+    respawn_interval_secs: u64,
+    shutdown: ShutdownSignal,
+    out: Pipeline,
+) -> crate::Result<super::Source> {
+    Ok(Box::pin(async move {
+        if respawn_on_exit {
+            let duration = time::Duration::from_secs(respawn_interval_secs);
+
+            // Continue to loop while shutdown is pending
+            while futures::poll!(shutdown.clone()).is_pending() {
```

Comment:
Can this and the one below use `poll!(&mut shutdown)` instead of cloning it each time? This is what the `generator` source does. The parameter would also need to be marked as `mut`, of course.

Reply:
This logic has now been changed to use `tokio::select!` per @jszwedko's suggestion. I tried to use `&mut shutdown` but with the new logic I could not think of a way to do that as `shutdown` was used in multiple places.